cilium/CFP-28986-mutual-auth-updates.md

```diff
@@ -0,0 +1,256 @@
+
+Here we have two nodes of a Cilium cluster, meeting the above requirements.
+
+The ipcaches are synced between the two nodes, so all the identities are correct.
+
+Both nodes `A` and `B` agree on the IP addresses associated with the identites `foo` and `bar`.
+
+Once the exploit has been performed, the nodes look more like this:
+
+![Node state after cache manipulation exploit](images/28986-after-exploit.png)
+
+On Node `B`, the connection to the apiserver has been interrupted in some way, so the Cilium agent is no longer receiving updates to Cilium Endpoints, meaning that the ipcache will become out of sync.
+
+On node `A`, enough pods have been scheduled in a new identity (the `baz` identity), to have the IP address that was previously associated with `foo` be now associated with `baz`.
+
+However, on Node `B`, because the ipcache is out of sync, the ingress policy will allow traffic from the pod that's actually a `baz` pod, because it has incorrect information about the IP to identity mapping. This is the reason why the exploit only works in direct routing mode - because in encapsulated modes, the identity is passed in any packet and used directly, rather than relying on the ipcache.
+
+This CFP lays out a proposal to mitigate this risk in two ways:
+* Require the use of either encryption (recommended because it prevents man-in-the-middle attacks) or encapsulation (which prevents cache manipulation but not MITM). In both of these cases, the Cilium identity is included in every packet, and so the ipcache is not used to impute identity to packets.
+* Add an additional mutual authentication mode that will use the connection tracking ("conntrack") table in the conntrack BPF map to store information about what _connections_ need authentication, instead of what _identity-identity-node pairings_ need authentication.
+
+The first idea makes cache manipulation irrelevant, because the inbound packets include the security identity in a header.
+
+However, in the existing mode, mutual authentication handshakes are only initiated when there is a change in either endpoint state (that is, a new node to node connection from a new pod), or when a certificate is rotated. So, in the event that an attacker _did_ manage to compromise mutual authentication mechanisms somehow, the window is, at maximum, the certificate rotation period, which defaults to 30 minutes.
+
+Adding a connection-based mutual authentication mode significantly reduces the timescale that an attack is exploitable for - since the mutual auth handshake is performed on each new connection, and connections _generally_ last much less that 30 minutes.
+
+It also makes performing the exploit harder, since the full 5-tuple (source IP, source port, destination IP, destination port, protocol) must match, rather than just source and destination IP, _and_ must also migrate the actual connection from one application to another. Still possible, but orders of magnitude harder.
```

Comment:
I don't find this very convincing. We know that the source and destination IP address match as that's a prerequisite for the attack. Matching the protocol is trivial. Is matching ports that difficult?

Reply:
It's not the destination port that's hard, it's actually the source port. The source port for connections is usually randomly chosen by the network stack, so simply starting a new connection will probably not work. The attacker will need to inspect existing flows and ensure that a new flow matches the existing source port.

In addition, the creation of a new connection will (for TCP connections) involve a new handshake, which should trigger the partially-closed parts of the datapath to clear the connection properties (including the mutual auth bits, as mentioned below).

Reply:
I added some explanation of the latter part.